internal/transport/http2_server.go

```diff
@@ -598,6 +599,22 @@ func (t *http2Server) operateHeaders(ctx context.Context, frame *http2.MetaHeade
 	if len(t.activeStreams) == 1 {
 		t.idle = time.Time{}
 	}
+	// Start a timer to close the stream on reaching the deadline.
+	if timeoutSet {
+		// We need to wait for s.cancel to be updated before calling
+		// t.closeStream to avoid data races.
+		cancelUpdated := make(chan struct{})
+		timer := internal.TimeAfterFunc(timeout, func() {
+			<-cancelUpdated
+			t.closeStream(s, true, http2.ErrCodeCancel, false)
+		})
+		oldCancel := s.cancel
+		s.cancel = func() {
+			oldCancel()
+			timer.Stop()
+		}
+		close(cancelUpdated)
```

Comment:
Does this work?  `oldCancel` could be called multiple times, but I think that's OK if it's just doing a context cancelation.

```suggestion
		oldCancel := s.cancel
		timer := internal.TimeAfterFunc(timeout, func() {
			oldCancel()
			t.closeStream(s, true, http2.ErrCodeCancel, false)
		})
		s.cancel = func() {
			timer.Stop()
			oldCancel()
		}
```

Reply:
The data race happens when reading and writing `s.cancel`. The timer fires, calling `t.closeStream` which in turn calls `s.cancel`.  The race is still possible with this change because the timer is started in line 605 and s.cancel is updated afterwards in line 609. 

<details>
<summary>Stack trace</summary>

```
WARNING: DATA RACE
Read at 0x00c000cc40e0 by goroutine 30741:
  google.golang.org/grpc/internal/transport.(*http2Server).closeStream()
      grpc-go/internal/transport/http2_server.g
o:1337 +0x44
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders.func4()
      grpc-go/internal/transport/http2_server.g
o:607 +0x57
  google.golang.org/grpc/internal/leakcheck.(*timerFactory).timeAfterFunc.func1()
      grpc-go/internal/leakcheck/leakcheck.go:2
92 +0x3c

Previous write at 0x00c000cc40e0 by goroutine 30738:
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      grpc-go/internal/transport/http2_server.g
o:609 +0x2e32
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      grpc-go/internal/transport/http2_server.g
o:680 +0x3fb
  google.golang.org/grpc.(*Server).serveStreams()
      grpc-go/server.go:1029 +0x67a
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      grpc-go/server.go:964 +0x86

Goroutine 30741 (running) created at:
  time.goFunc()
      /usr/lib/google-golang/src/time/sleep.go:215 +0x44

Goroutine 30738 (running) created at:
  google.golang.org/grpc.(*Server).handleRawConn()
      grpc-go/server.go:963 +0x2e9
  google.golang.org/grpc.(*Server).Serve.func3()
      grpc-go/server.go:932 +0x7b
```
</details>

Running all the tests a couple of times should surface the failure. One way to avoid synchronization is to pass the cancellation function to `closeStream` as a parameter. The closure executed by the timee can pass `oldCancel`, existing callers can pass `s.cancel`.